packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
+    }
+  };
 
-      if (ev.target !== ev.currentTarget && !(ev.target === firstBumper.current || ev.target === lastBumper.current)) {
-        // every time focus changes within the trap zone, remember the focused element so that
-        // it can be restored if focus leaves the pane and returns via keystroke (i.e. via a call to this.focus(true))
-        internalState.previouslyFocusedElementInTrapZone = ev.target as HTMLElement;
-      }
-    },
-    [onFocusCapture, internalState, onBumperFocus],
-  );
+  /** Called to restore focus on unmount or props change. (useEventCallback ensures latest prop values are used.) */
+  const returnFocusToInitiator = useEventCallback((elementToFocusOnDismiss: HTMLElement | null) => {
+    FocusTrapZone.focusStack = FocusTrapZone.focusStack!.filter(value => internalState.focusStackId !== value);
 
-  const returnFocusToInitiator = React.useCallback((): void => {
-    FocusTrapZone.focusStack = FocusTrapZone.focusStack.filter((value: any) => {
-      return id !== value;
-    });
-
-    if (doc) {
-      const activeElement = doc.activeElement as HTMLElement;
-      if (
-        !ignoreExternalFocusing &&
-        internalState.previouslyFocusedElementOutsideTrapZone &&
-        typeof internalState.previouslyFocusedElementOutsideTrapZone.focus === 'function' &&
-        (elementContains(root.current, activeElement) || activeElement === doc.body)
-      ) {
-        if (
-          !(
-            internalState.previouslyFocusedElementOutsideTrapZone === firstBumper.current ||
-            internalState.previouslyFocusedElementOutsideTrapZone === lastBumper.current
-          )
-        ) {
-          focusAsync(internalState.previouslyFocusedElementOutsideTrapZone);
-        }
-      }
+    if (!doc) {
+      return;
     }
-  }, [doc, id, ignoreExternalFocusing, internalState]);
 
-  const forceFocusInTrap = React.useCallback(
-    (ev: FocusEvent): void => {
-      if (disabled) {
-        return;
-      }
-      if (FocusTrapZone.focusStack.length && id === FocusTrapZone.focusStack[FocusTrapZone.focusStack.length - 1]) {
-        const focusedElement = ev.target as HTMLElement;
-        if (!elementContains(root.current, focusedElement)) {
-          focus();
-          internalState.hasFocus = true; // set focus here since we stop event propagation
-          ev.preventDefault();
-          ev.stopPropagation();
-        }
-      }
-    },
-    [disabled, id, focus, internalState],
-  );
+    const activeElement = doc.activeElement as HTMLElement;
+    if (
+      !disableRestoreFocus &&
+      typeof elementToFocusOnDismiss?.focus === 'function' &&
+      // only restore focus if the current focused element is within the FTZ, or if nothing is focused
+      (elementContains(root.current, activeElement) || activeElement === doc.body)
+    ) {
+      focusElementAsync(elementToFocusOnDismiss);
+    }
+  });
 
-  const forceClickInTrap = React.useCallback(
```

Comment:
This has been combined with `forceFocusInTrap` because they were identical aside from the event type (which doesn't matter in this context)